libs/constrained-generators/src/Constrained/Properties.hs

```diff
@@ -69,20 +91,130 @@ prop_conformEmpty ::
   QC.Property
 prop_conformEmpty a = QC.property $ conformsTo @fn a (emptySpec @fn @a)
 
-prop_propagateSpec ::
-  (HasSpec fn a, HasSpec fn b) =>
-  Var b ->
-  Term fn a ->
-  Specification fn a ->
-  QC.Property
-prop_propagateSpec var tm spec =
-  let spec' = errorGE $ do
-        ctx <- toCtx var tm
-        pure $ propagateSpec spec ctx
-   in QC.forAll (strictGen $ genFromSpec spec') $ \geval -> fromGE (\err -> QC.counterexample (unlines err) False) $ do
-        val <- geval
-        res <- runTerm (singletonEnv var val) tm
-        pure $ QC.property $ conformsToSpec res spec
+prop_univSound :: forall fn. TestableFn fn -> QC.Property
+prop_univSound (TestableFn fn) =
+  QC.label (show fn) $
+    QC.forAllShrinkBlind QC.arbitrary QC.shrink $ \(TestableCtx ctx :: TestableCtx fn as) ->
+      QC.forAllShrinkBlind QC.arbitrary QC.shrink $ \(spec :: Specification fn a) ->
+        QC.counterexample ("\nfn ctx = " ++ showCtxWith fn (TestableCtx ctx)) $
+          QC.counterexample (show $ "\nspec =" <+> pretty spec) $
+            let sspec = simplifySpec (propagateSpecFun fn ctx spec)
+             in QC.counterexample ("\n" ++ show ("propagateSpecFun fn ctx spec =" /> pretty sspec)) $
+                  QC.forAllBlind (strictGen $ genFromSpec @_ @_ @GE sspec) $ \ge ->
+                    fromGEDiscard $ do
+                      a <- ge
+                      let res = uncurryList_ unValue (sem fn) $ fillListCtx ctx $ \HOLE -> Value a
+                      pure $
+                        QC.counterexample ("\ngenerated value: a = " ++ show a) $
+                          QC.counterexample ("\nfn ctx[a] = " ++ show res) $
+                            conformsToSpecProp res spec
```

Comment:
I think it would make sense to create a monad that can get rid of this staircasing, since this pattern appears quite often.

Reply:
It would indeed, but thats a QuickCheck problem I think rather than something to worry about here. I'm more inclined in this case to say that it's a formatter issue to be honest.

Reply:
Thank god for the formatter, otherwise it would be unreadable :stuck_out_tongue_winking_eye: 

In any case, there is no need to complicate the code any more than necessary with extra monads. There is nothing wrong with the indentation and "staircasing"